### DISCUSSION TITLE
Importing p12 files with password in Match

### DISCUSSION THREAD

**Discussion by iOSGeekster** [fastlane/fastlane#19814] [Q&A]
Hi!

Is it correctly understood that match doens't support importing p12 files with a password yet?

To me it seems like the `Utils.import` in the `runner.rb` doesn't pass on `CERTIFICATE_PASSWORD` environment variable to the `FastlaneCore::KeychainImport.import_file` method. 

So I was thinking if the `match import` should, or could, take a `certificate_password` argument, that would be passed down?

**Comment by maklostermann** [comment]
Hello,

I stumbled upon this discussion as I was trying to use match import. After struggling with the import for quite some time, I found out that the key file must be an unencrypted file. At least on MacOS, an encrypted PKCS file cannot be imported with an empty password.

As match encrypts the key file anyway, using an encrypted file with an additional password is not important to me.

However, it took me quite some time to find out how it works (reading the code and analyzing an existing match repository). It would be nice if the documentation (website and command options) of match import would be improved:
- Certificate must be DER encoded (as far as I know).
- Key must be PKCS in PEM (unencrypted).

It would be even nicer if the import command would verify the supplied files before adding them to the repository.

Another limitation is, that it will only work if the certificate and key file have the same name (if I didn't miss something in the code).

**Reply by nicobao** [reply]
I agree with this take, I just spent lots of time before realizing that was the problem!

**Reply by wow-such-amazing** [reply]
I tried to import pem file for private key, but match failed during the build with this error:
```
security: SecKeychainItemImport: Unknown format in import.
```

And here is our code to extract certificate with private key and import them:
```
echo "➡️  Extract the Cert.pem from the Archive.p12 file"

openssl pkcs12 \
-nodes \
-legacy \
-nokeys \
-clcerts \
-in $archivePath \
-passin pass:$(< $archivePasswordPath) \
-out $certPemPath

echo "➡️  Extract the Key.pem from the Archive.p12 file"

openssl pkcs12 \
-nodes \
-legacy \
-nocerts \
-passin pass:$(< $archivePasswordPath) \
-in $archivePath \
-out $keyPemPath


echo "➡️  Convert the Cert.pem to Cert.cer file"

openssl x509 \
-outform der \
-in $certPemPath \
-out $certCerPath

echo "➡️  Verify the Cert.cer file"

openssl x509 \
-inform der \
-text \
-in $certCerPath \
-noout

echo "➡️  Verify the Cert + Key Pair"

openssl x509 -noout -modulus -in $certPemPath | openssl md5
openssl rsa -noout -modulus -in $keyPemPath | openssl md5
```

**Comment by siegeral** [comment]
Is there any progress regarding this feature request?

**Comment by Lutzifer** [comment]
As I struggled for a long time today to do this (knowing I successfully did this before):

To move an existing Enterprise Certificate/Key Pair from the Keychain Access App to match do the following: 

1. Open Keychain Access
2. Go into the certificates tab
3. Filter for the certificate you want to export
4. Klick the small “>” next to the certificate to reveal the key
5. Right Klick on the key, choose export
6. Choose the “P12” Format, klick “OK”
7. Do not set a password, just klick “OK”
8. Right Click on the certificate, choose export
9. Choose the “CER” Format, klick “OK”
10. Copy the Certificates.cer and Certificates.p12 files into your project folder
11. Run

```
bundle exec fastlane match import --readonly true --type enterprise --skip_certificate_matching true
```

When asked for the .cer file, write Certificates.cer
When asked for the .p12 file, write Certificates.p12
 When asked for the mobileprovisioning files, just hit enter.
 
 To import a profile, run e.g.

```
openssl aes-256-cbc -k <MATCH_PASSWORD> -in unencrypted.mobileprovision -out encrypted.mobileprovision -a
```

and upload the result to the "profiles/enterprise" folder of your match repo.

**Reply by Lutzifer** [reply]
updated the docs in https://github.com/fastlane/fastlane/pull/22176

**Comment by wow-such-amazing** [comment]
Would love to have a way to provide a password during match import or to have more docs on how to remove a password from a certificate. It's pretty confusing on how to properly do that. 

One confusing part is that my `p12` file contains both Certificate and Private Key. So I'm not sure if I need a separate `Cert.cer` file. I'm also unsure if new `p12` file should contain only Private Key or if I should include Certificate as well.

Then @maklostermann mentioned that it should be non-encrypted `pem` file instead of `p12` for the private key. 

So much confusion and unsure where to find my answer 😕

**Reply by wow-such-amazing** [reply]
@joshdholtz sorry to ping you, but maybe you know something about this topic or know someone who could shed some light 🙏

**Reply by wow-such-amazing** [reply]
Here is my current attempt. I tried multiple different setting at the moment, but non worked. Match import works fine, then match fetch during the build works as well. But then build fails when it tried to do `CodeSign` 😢

```
echo "➡️  Extract the Cert.pem from the Archive.p12 file"

openssl pkcs12 \
-nodes \
-legacy \
-nokeys \
-clcerts \
-in $archivePath \
-passin pass:$(< $archivePasswordPath) \
-out $certPemPath

echo "➡️  Extract the Key.pem from the Archive.p12 file"

openssl pkcs12 \
-nodes \
-legacy \
-nocerts \
-passin pass:$(< $archivePasswordPath) \
-in $archivePath \
-out $keyPemPath

echo "➡️  Convert the Cert.pem to Cert.cer file"

openssl x509 \
-outform der \
-in $certPemPath \
-out $certCerPath

echo "➡️  Verify the Cert.cer file"

openssl x509 \
-inform der \
-text \
-in $certCerPath \
-noout

# echo "➡️  Convert the Key.pem to Key.p12 file"

openssl pkcs12 \
-export \
-nocerts \
-passout pass: \
-inkey $keyPemPath \
-out $keyP12Path

# echo "➡️  Verify the Key.p12 file"

openssl pkcs12 -info -passin pass: -in $keyP12Path -noout

echo "➡️  Verify the Cert + Key Pair"

openssl x509 -noout -modulus -in $certPemPath | openssl md5
openssl rsa -noout -modulus -in $keyPemPath | openssl md5
```

**Reply by wow-such-amazing** [reply]
I have also tried what @Lutzifer wrote down. I first imported files into my Keychain, and then exported again without a password. Uploaded to our repo successfully:
```
[11:38:23]: Cloning remote git repo...
[11:38:26]: Add git user config to local git repo...
[11:38:26]: Checking out branch TEAM_ID...
[11:38:26]: 🔓  Successfully decrypted certificates repo
[11:38:26]: Repo is at: '/var/folders/BlaBlaBla'
[11:38:26]: Login to App Store Connect (email@email.com)
Available session is not valid anymore. Continuing with normal login.
[11:38:33]: 🔒  Successfully encrypted certificates repo
[11:38:33]: Pushing changes to remote git repo...
[11:38:35]: Finished uploading files to Git Repo [git@github.com:BlaBla-Corporation/BlaBlaRepo.git]
```
But the build has failed again with the same error during CodeSign 😕 
```
[10:52:39]: ▸ [35m** ARCHIVE FAILED **[0m
[10:52:39]: ▸ [35mThe following build commands failed:[0m
[10:52:39]: ▸ [35m	CodeSign /Users/jenkins/workspace/iOS_master/DerivedData/Build/Intermediates.noindex/ArchiveIntermediates/Swapcard/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/ZIPFoundation_ZIPFoundation.bundle (in target 'ZIPFoundation_ZIPFoundation' from project 'ZIPFoundation')[0m
```

**Reply by Lutzifer** [reply]
> I have also tried what @Lutzifer wrote down. I first imported files into my Keychain, and then exported again without a password. Uploaded to our repo successfully:
> 
> ```
> [11:38:23]: Cloning remote git repo...
> [11:38:26]: Add git user config to local git repo...
> [11:38:26]: Checking out branch TEAM_ID...
> [11:38:26]: 🔓  Successfully decrypted certificates repo
> [11:38:26]: Repo is at: '/var/folders/BlaBlaBla'
> [11:38:26]: Login to App Store Connect (email@email.com)
> Available session is not valid anymore. Continuing with normal login.
> [11:38:33]: 🔒  Successfully encrypted certificates repo
> [11:38:33]: Pushing changes to remote git repo...
> [11:38:35]: Finished uploading files to Git Repo [git@github.com:BlaBla-Corporation/BlaBlaRepo.git]
> ```
> 
> But the build has failed again with the same error during CodeSign 😕
> 
> ```
> [10:52:39]: ▸ �[35m** ARCHIVE FAILED **�[0m
> [10:52:39]: ▸ �[35mThe following build commands failed:�[0m
> [10:52:39]: ▸ �[35m	CodeSign /Users/jenkins/workspace/iOS_master/DerivedData/Build/Intermediates.noindex/ArchiveIntermediates/Swapcard/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/ZIPFoundation_ZIPFoundation.bundle (in target 'ZIPFoundation_ZIPFoundation' from project 'ZIPFoundation')�[0m
> ```

Make sure to delete any old profiles and certs in the match repo before uploading new ones, they are not replaced in place.

**Reply by wow-such-amazing** [reply]
@Lutzifer that's a good point! I figured this as well during my certificates journey and now we have this command before we run match during the build 👀 

It finds all certificates that contain a team id in the name and removes them from keychain
```
sh "security find-certificate -c \"#{$portal.adc_team_id}\" -a -Z | sudo awk '/SHA-1/{system(\"security delete-certificate -Z \"$NF)}'"
```

**Reply by MihailStoicaGH** [reply]
@wow-such-amazing did you find the solution? I have the same problem.

**Reply by wow-such-amazing** [reply]
@MihailStoicaGH we were able to import certificates couple times manually, but I still find it quite fragile. Mostly because we don't have control over how those certificates are being created and exported. And hence in case of an error it's hard to tell if it's on our end or on the end of the person who generated a certificate.

So most of the time we try to avoid importing certificates if possible.

As for the issue in this thread I don't recall a proper solution. Tried multiple options until something has worked 😅 Maybe also certificate was re-generated or re-exported 👀

**Reply by MihailStoicaGH** [reply]
My main issue is that the import works, the fetch works but it can't sign the build, error 65 while archiving. This is through gitlab ci which runs a fastlane lane.
If I manually run that lane on my local machine, with an empty keychain, the signing works.

**Reply by wow-such-amazing** [reply]
@MihailStoicaGH that looks to me like an older certificate is being used on CI from keychain. As I mentioned I've added a step to clean up keychain from same team certificates so that there is no collision

**Reply by MihailStoicaGH** [reply]
@wow-such-amazing I'm using a custom keychain and before running any lane I'm deleting the old one, re-creating it and the unlock it. The certificates are fine. My guess is that gitlab secure files isn't a good approach when using fastlane match.

**Comment by rgkobashi** [comment]
I tried to remove the password from the p12 using so many ways and I was still getting error when signing builds.
Turns out, the problem was not on the exporting/importing, it was on the trust settings of the certificate:

```
error: Invalid trust settings. Restore system default trust settings for certificate ... in order to sign code with it
```

And the way to fix it:

1. Locate the certificate on the keychain, right click on it and then `Get Info`
2. Expand the `Trust` section and make sure `Use System Defaults` is selected on `When using this certificate` (using `Always Trust` does not work)

**Comment by vnagendra** [comment]
I spent a considerable amount of time trying to decode and understand this. Hopefully the following writeup helps someone else not waste a day on it.

There are a couple of things about Developer ID Installer / Application certs with `fastlane match` that aren't obvious.

## Generating the cert is the same for both

The generation is _exactly_ the same on both counts:

- Generate an OpenSSL key and then a CSR.
- Upload the CSR to Apple and go through the manual flow to download a `.cer` (the signed certificate).
- The names don't matter much. As a point of reference I used `developer_id_installer.{key,csr,p12}`, and Apple hands you back `developerID_installer.cer`.

What matters is **what you do with that `.cer`**. Either way you need to bundle a `.p12` to import into match — you're most likely doing this to distribute to multiple developers / CI pipelines. So ultimately you need two artifacts: `developer_id_installer.p12` and `developerID_installer.cer`.

## Problem 1 — the blank-password p12 (both cert types)

Normal thinking is: take the `.cer` and `.key` and use OpenSSL to export a `.p12`. The catch is that the p12 must have a **blank** password, because match imports it with a hardcoded `security import -P ""` (there's no option to pass a p12 password). So you do:

```bash
openssl pkcs12 -export ... -passout pass:    # blank password
```

...and on a current macOS it fails at import time with:

```
security: SecKeychainItemImport: MAC verification failed during PKCS12 import (wrong password?)
```

> **This is the core problem.** OpenSSL and Apple compute the PKCS#12 integrity MAC over an _empty_ (`""`) password differently — they disagree on how a zero-length password is encoded going into the key derivation. So a blank-password p12 built by OpenSSL (or LibreSSL) produces a MAC that Apple's `security import` can't verify, and you get the misleading "wrong password?" error even though the password really is blank. OpenSSL does not give you a flag for encoding blank password the "apple way".

The fix is to let **Apple** write the blank-password p12 instead of OpenSSL:

1. Bundle the `.cer` + `.key` into a p12 **with** a throwaway password (Apple's keychain will import that fine).
2. Import that p12 into a keychain.
3. Export it back out of the keychain with a **blank** password.

That export carries Apple's own empty-password encoding, so `match`'s `-P ""` import accepts it everywhere. Here's a self-contained version that does it in a throwaway keychain so it never touches your login keychain:

```bash
# Key + CSR
openssl genrsa -out developer_id_installer.key 2048
openssl req -new -key developer_id_installer.key \
    -out developer_id_installer.csr \
    -subj "/emailAddress=you@example.com/CN=Your Company/C=US"

# >>> MANUAL, in the browser <<<
#   developer.apple.com → Certificates → +  →  pick "Developer ID Installer"
#   (NOT "Apple Distribution", NOT "Mac Installer Distribution").
#   Upload developer_id_installer.csr, download the cert, save it as:
#       developerID_installer.cer     (DER, the way Apple hands it to you)
openssl x509 -inform DER -in developerID_installer.cer -out developer_id_installer.pem

# Sanity check before bundling:
#   (a) subject MUST say "Developer ID Installer:"
openssl x509 -in developer_id_installer.pem -noout -subject
#   (b) cert and key must be the same pair — these two hashes must match
openssl x509 -in developer_id_installer.pem -noout -modulus | openssl md5
openssl rsa  -in developer_id_installer.key -noout -modulus | openssl md5

# Bundle the p12, laundered through a throwaway keychain so it carries Apple's
# empty-password MAC. The result opens with a blank passphrase.
KC="$PWD/launder.keychain-db"
security create-keychain -p temp "$KC"
security unlock-keychain  -p temp "$KC"
openssl pkcs12 -export -legacy \
    -inkey developer_id_installer.key \
    -in    developer_id_installer.pem \
    -name  "Developer ID Installer" \
    -passout pass:temp -out _pw.p12
security import _pw.p12 -k "$KC" -P temp -A
security export -k "$KC" -t identities -f pkcs12 -P "" -o developer_id_installer.p12
security delete-keychain "$KC"
rm -f _pw.p12

# Verify: opens with a blank passphrase AND contains BOTH a cert and a key.
# You want to see a "Certificate bag".
openssl pkcs12 -info -in developer_id_installer.p12 -nodes -passin pass: -noout
```

> Note `security export -t identities` exports **every** identity in the keychain — that's exactly why this uses a fresh throwaway `launder.keychain-db` and deletes it right after. Don't point it at your login keychain.

The Developer ID Application cert is the same recipe with different names.

## Problem 2 — `match import` for Developer ID Installer hits the Connect API (installer only)

When you go to import the installer cert:

```bash
fastlane match import --type developer_id_installer
```

...it crashes with something like:

```
'DEVELOPER_ID_INSTALLER' is not a valid filter value. Expected one of: ...
MAC_INSTALLER_DISTRIBUTION, ... DEVELOPER_ID_APPLICATION, DEVELOPER_ID_APPLICATION_G2, ...
```

What's happening: before storing your files, match calls the App Store Connect API to look up the certificate's ID (so it can name the stored file `<certID>.cer`). For `developer_id_installer` it filters on `certificateType=DEVELOPER_ID_INSTALLER` — but Apple's Connect API doesn't expose that as a queryable type. Its enum has `DEVELOPER_ID_APPLICATION`, `DEVELOPER_ID_APPLICATION_G2`, `MAC_INSTALLER_DISTRIBUTION`, `DEVELOPER_ID_KEXT`… but **no `DEVELOPER_ID_INSTALLER`**. So there's no value match could send that would work — Apple lets you _have_ the cert but won't let you _list_ it on that endpoint.

The fix is to skip that lookup entirely:

```bash
fastlane match import --type developer_id_installer --skip_certificate_matching true
#   certificate path : developerID_installer.cer     (the DER from Apple)
#   private key path : developer_id_installer.p12     (laundered, from above)
```

With `--skip_certificate_matching true`, match never calls the API — it just stores your files using their basenames. The only thing you give up is the cert-ID-based filename, which doesn't matter. (If you're replacing an existing cert, remove the old `certs/developer_id_installer/<oldID>.{cer,p12}` from the repo first so the slot holds exactly one pair.)

The two basenames differing (camelCase `.cer`, snake_case `.p12`) is fine — a readonly install globs `*.cer` and `*.p12` in the slot independently and installs the single pair regardless of name.

The Developer ID **Application** import does _not_ need this flag — `developer_id_application` maps to `DEVELOPER_ID_APPLICATION(_G2)`, which the API does recognize.

## What if you don't see the installer under `-p codesigning`

After a successful `fastlane match --readonly`, you might check with:

```bash
security find-identity -v -p codesigning
```

...and only see the **Application** cert, and worry that the installer didn't import. It did. A Developer ID **Installer** cert doesn't sign code — it signs installer packages (`.pkg`) via `productsign`/`productbuild` — so it's deliberately excluded from the `codesigning` policy. Drop the `-p` filter to see everything:

```bash
security find-identity -v | grep "Developer ID"
```

Both `Developer ID Application` and `Developer ID Installer` should be there.
